### DISCUSSION TITLE
Recommended directory structure for extension development?

### DISCUSSION THREAD

**Discussion by kevcao-certik** [microsoft/vscode-discussions#230] [Extension Development QnA] [answered]
What is the typical recommended directory structure when making a VSCode extension?

From what I gather after looking at a couple of open-source extensions, within `src/`, there is usually a `common/` folder to hold utility/library files shared across the extension. What if your extension has multiple features? How do you separate the front-end components of a feature from the back-end?

**Comment by bwateratmsft** [comment, marked as answer]
The easiest starting point is to use Yeoman's VSCode extension template generator.

```sh
npm install -g yo generator-code
yo code
```

Unless your extension is extremely simple, I would highly recommend saying "yes" to the question of whether or not to bundle the code with webpack.

Generally extensions will have a `src/` folder, and an `extension.ts` file within it, that acts as the main entry point for the `activate()` and `deactivate()` methods that all extensions require. Beyond that, how you structure things is entirely up to you. The Yeoman template gives an excellent starting point that includes webpacking, the necessary VSCode launch configurations to debug and run tests, and other nice things.
